DIG-1582: Implement overview thresholds & DIG-1533: Change cancer_type to primary_site

SonQBChau opened the pull request:
## What's new
- Introduce a threshold mechanism to display masked values for small counts, e.g., "<5".
- Rewrite queries for masking and improved performance.
- Update return types from int to str to handle masked values.
- Update schema return structure to accommodate new queries.
- Rename `discover_cancer_type_count` to `discover_primary_site_count`
- Use lazyFunction for enhanced random seeds in testing
- Add a test suite to ensure proper functionality of overview masking for small counts.

### Types of Change(s)
- [x] ✨ New feature 
- [x] 💥 Breaking change

Which other services does it affect?
- [x] data-portal
- [x] query

### Has it been tested for:
- [x] Locally tested
- [x] Docker tested

### How to test
- For local, run tox
- For docker, pull this branch, add to docker-compose ` AGGREGATE_COUNT_THRESHOLD=${AGGREGATE_COUNT_THRESHOLD}` and run test-integration
- Optional: Add some data and call overview endpoints to see if value <5 gets properly masked.

SonQBChau committed:
Update discovery.py

- Introduced threshold for small counts to display masked values, e.g., <5.
- Rewrite queries for masking and performance.
- Updated return types from int to str to cope with "<5"
- Update schema return structure for new queries

SonQBChau committed:
add new schema for overview masking count

SonQBChau committed:
use lazyFunction for better random seeds

SonQBChau committed:
add test suite for overview masking small count

SonQBChau committed:
add AGGREGATE_COUNT_THRESHOLD var

SonQBChau committed:
update schemas

SonQBChau committed:
Update schemas with SHA

SonQBChau committed:
bump katsu version

SonQBChau committed:
Merge branch 'sonchau/overview_threshold' of https://github.com/CanDIG/katsu into sonchau/overview_threshold

SonQBChau committed:
clean up

CourtneyGosselin reviewed (approved):
Approved! I was able to spin up the environment and test some of the calls. The results return as expected, showing the new AGGREGATE_COUNT_THRESHOLD <5

OrdiNeu commented:
Getting an error on the frontpage with this PR, from Katsu's logs:
```
Cannot use None as a query value                                                                                                                                                                   [22/1852]
Traceback (most recent call last):                                                                                                                                                                          
  File "/usr/local/lib/python3.12/site-packages/ninja/operation.py", line 107, in run
    result = self.view_func(request, **values)   
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                              
  File "/app/chord_metadata_service/chord_metadata_service/mohpackets/apis/discovery.py", line 302, in discover_diagnosis_age_count
    .annotate(                   
     ^^^^^^^^^                                                                                        
  File "/usr/local/lib/python3.12/site-packages/django/db/models/query.py", line 1630, in annotate
    return self._annotate(args, kwargs, select=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                        
  File "/usr/local/lib/python3.12/site-packages/django/db/models/query.py", line 1680, in _annotate
    clone.query.add_annotation(                                                                       
  File "/usr/local/lib/python3.12/site-packages/django/db/models/sql/query.py", line 1198, in add_annotation
    annotation = annotation.resolve_expression(self, allow_joins=True, reuse=None)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/expressions.py", line 1513, in resolve_expression
    c.cases[pos] = case.resolve_expression(                                                                                                                                                                 
                   ^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                   File "/usr/local/lib/python3.12/site-packages/django/db/models/expressions.py", line 1432, in resolve_expression
    c.condition = c.condition.resolve_expression(                                                                                                                                                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/query_utils.py", line 90, in resolve_expression                                                                               
    clause, joins = query._add_q(                                                                     
                    ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/sql/query.py", line 1645, in _add_q
    child_clause, needed_inner = self.build_filter( 
                                 ^^^^^^^^^^^^^^^^^^ 
  File "/usr/local/lib/python3.12/site-packages/django/db/models/sql/query.py", line 1513, in build_filter
    condition = self.build_lookup(lookups, reffed_expression, value)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/sql/query.py", line 1394, in build_lookup
    raise ValueError("Cannot use None as a query value")
ValueError: Cannot use None as a query value
[21/May/2024 16:06:46] [django] ERROR: Cannot use None as a query value
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/ninja/operation.py", line 107, in run
    result = self.view_func(request, **values)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/chord_metadata_service/chord_metadata_service/mohpackets/apis/discovery.py", line 302, in discover_diagnosis_age_count
    .annotate(
     ^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/query.py", line 1630, in annotate
    return self._annotate(args, kwargs, select=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/query.py", line 1680, in _annotate
    clone.query.add_annotation(
  File "/usr/local/lib/python3.12/site-packages/django/db/models/sql/query.py", line 1198, in add_annotation
    annotation = annotation.resolve_expression(self, allow_joins=True, reuse=None)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/expressions.py", line 1513, in resolve_expression
    c.cases[pos] = case.resolve_expression(
                   ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/expressions.py", line 1432, in resolve_expression
    c.condition = c.condition.resolve_expression(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/query_utils.py", line 90, in resolve_expression
    clause, joins = query._add_q(
                    ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/sql/query.py", line 1645, in _add_q
    child_clause, needed_inner = self.build_filter(
                                 ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/sql/query.py", line 1513, in build_filter
    condition = self.build_lookup(lookups, reffed_expression, value)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/sql/query.py", line 1394, in build_lookup
    raise ValueError("Cannot use None as a query value")
ValueError: Cannot use None as a query value
Internal Server Error: /v2/discovery/overview/diagnosis_age_count/
[21/May/2024 16:06:46] [django.request] ERROR: Internal Server Error: /v2/discovery/overview/diagnosis_age_count/
[pid: 16|app: 0|req: 43/43] 172.20.0.1 () {32 vars in 2040 bytes} [Tue May 21 16:06:46 2024] GET /v2/discovery/overview/diagnosis_age_count/ => generated 2247 bytes in 10 msecs (HTTP/1.1 500) 9 headers in 664 bytes (1 switches on core 0)

```

SonQBChau commented:
I can confirm this bug, looking into it

SonQBChau committed:
fix None value not a string

SonQBChau commented:
fixed. Can you check it again @OrdiNeu ?

OrdiNeu commented:
I've rebuilt a few times with the latest changes but still get the same error listed above. For reference, I'm just visiting the frontpage and checking the logs -- it's a request straight to Katsu's `/v2/discovery/overview/diagnosis_age_count/` endpoint.

SonQBChau committed:
make required env check

SonQBChau commented:
@OrdiNeu Have you added ` AGGREGATE_COUNT_THRESHOLD=${AGGREGATE_COUNT_THRESHOLD}` to docker compose? (I put a note in the description, but even I forgot it myself).

To avoid this in the future, I have added the code to check if any .env is missing.

Let me know if it works for you.

SonQBChau committed:
Merge branch 'candig_develop' into sonchau/overview_threshold

SonQBChau committed:
clean up

OrdiNeu commented:
Oh, that sort of works now. Getting a new error now though, on `individual_count` only:
```
'>=' not supported between instances of 'int' and 'str'
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/ninja/operation.py", line 107, in run
    result = self.view_func(request, **values)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/chord_metadata_service/chord_metadata_service/mohpackets/apis/discovery.py", line 188, in discover_individual_count
    if donor_count >= SMALL_NUMBER_THRESHOLD
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: '>=' not supported between instances of 'int' and 'str'
[22/May/2024 14:56:51] [django] ERROR: '>=' not supported between instances of 'int' and 'str'
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/ninja/operation.py", line 107, in run
    result = self.view_func(request, **values)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/chord_metadata_service/chord_metadata_service/mohpackets/apis/discovery.py", line 188, in discover_individual_count
    if donor_count >= SMALL_NUMBER_THRESHOLD
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: '>=' not supported between instances of 'int' and 'str'
Internal Server Error: /v2/discovery/overview/individual_count/
[22/May/2024 14:56:51] [django.request] ERROR: Internal Server Error: /v2/discovery/overview/individual_count/
```

SonQBChau committed:
fix discover_individual_count

- fix the bug where AGGREGATE_COUNT_THRESHOLD read as a str in docker env
- should show 0 instead of <5 for value 0

SonQBChau commented:
I fixed the bug where AGGREGATE_COUNT_THRESHOLD is not an int. Accidentally, I found a bug with cache but I will address it in another ticket

OrdiNeu commented on config/settings/prod.py:
```diff
@@ -10,13 +10,34 @@
+    "REDIS_PASSWORD_FILE",
+]
+
+missing_vars = [var for var in required_env_vars if not os.getenv(var)]
```
This file switches between using `os.getenv()` and `os.environ[]` in several places -- I'd recommend sticking to one or the other (as far as I understand it, one's a wrapper of the other)

SonQBChau replied:
There is a slight difference, and it's on purpose. `os.getenv` returns None if the variable is not found, allowing me to catch it and throw an exception.

Why not use `os.getenv` below? If it returns None for a missing variable, it can cause the bug we encountered before. If I (or someone else) tries to get a new variable but forgets to add it to required_vars, it will crash instead of causing a subtle bug, making it easier to catch.

OrdiNeu replied:
I see -- perhaps it would be best if lines 34-36 used `os.getenv` then, rather than `os.environ.get()`? For consistency

SonQBChau replied:
oh thank you for the eagle eye! There was a merge conflict and I used the upstream code without realized it

OrdiNeu commented on chord_metadata_service/mohpackets/tests/endpoints/test_overview.py:
```diff
@@ -0,0 +1,203 @@
+"""
+
+
+class OverviewTestCase(BaseTestCase):
```
These tests are well-written :+1:

OrdiNeu commented on chord_metadata_service/mohpackets/tests/endpoints/factories.py:
```diff
@@ -73,16 +73,20 @@ class Meta:
-    }
+    date_of_birth = factory.LazyFunction(
+        lambda: {
+            "month_interval": random.randint(0, 1000),
```
Wouldn't this make the month_interval and day_interval not line up? E.g. you could have a day_interval of 0 and a month_interval of 1000.

SonQBChau replied:
You are correct, but I'm not testing the date logic here. I agree that a more realistic generator would be better, but these mock data serve my purpose for now.

OrdiNeu replied:
Sounds good to me -- maybe document this discrepancy in the Note: at the top of this file. Otherwise it looks good :heavy_check_mark:

SonQBChau replied:
done

OrdiNeu reviewed (approved):
Looks good, well-written :heavy_check_mark: 

The latest fix made it work for me

OrdiNeu commented:
NB: we should probably hold on updating the submodule version on CanDIGv2's `develop` until we have both [Query PR](https://github.com/CanDIG/candigv2-query/pull/37) and the [Data Portal PR](https://github.com/CanDIG/candig-data-portal/pull/144) merged at the same time. In addition, we need a second Data Portal PR to handle censoring.

SonQBChau committed:
Update prod.py

- don't use get

SonQBChau committed:
Update factories.py
